kube-client/src/api/entry.rs

```diff
@@ -0,0 +1,161 @@
+#![warn(missing_docs)]
+
+use std::fmt::Debug;
+
+use kube_core::{params::PostParams, Resource};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{Api, Result};
+
+impl<K: Resource + Clone + DeserializeOwned + Debug> Api<K> {
+    pub async fn entry<'a>(&'a self, name: &'a str) -> Result<Entry<'a, K>> {
+        Ok(match self.try_get(name).await? {
+            Some(object) => Entry::Occupied(OccupiedEntry {
+                api: self,
+                object,
+                dirtiness: Dirtiness::Clean,
+            }),
+            None => Entry::Vacant(VacantEntry { api: self, name }),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Entry<'a, K> {
+    Occupied(OccupiedEntry<'a, K>),
+    Vacant(VacantEntry<'a, K>),
+}
+
+impl<'a, K> Entry<'a, K> {
+    pub fn get(&self) -> Option<&K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    pub fn get_mut(&mut self) -> Option<&mut K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get_mut()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    pub fn and_modify(self, f: impl FnOnce(&mut K)) -> Self {
+        match self {
+            Entry::Occupied(entry) => Entry::Occupied(entry.and_modify(f)),
+            entry @ Entry::Vacant(_) => entry,
+        }
+    }
+
+    pub fn or_insert(self, default: impl FnOnce() -> K) -> OccupiedEntry<'a, K>
+    where
+        K: Resource,
+    {
+        match self {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(entry) => entry.insert(default()),
+        }
+    }
+}
+
+pub struct OccupiedEntry<'a, K> {
+    api: &'a Api<K>,
+    dirtiness: Dirtiness,
+    object: K,
+}
+
+#[derive(Debug)]
+enum Dirtiness {
+    Clean,
+    Dirty,
+    New,
+}
+
+impl<'a, K> OccupiedEntry<'a, K> {
+    pub fn get(&self) -> &K {
+        &self.object
+    }
+
+    pub fn get_mut(&mut self) -> &mut K {
+        self.dirtiness = match self.dirtiness {
+            Dirtiness::Clean => Dirtiness::Dirty,
+            Dirtiness::Dirty => Dirtiness::Dirty,
+            Dirtiness::New => Dirtiness::New,
+        };
+        &mut self.object
+    }
+
+    pub fn and_modify(mut self, f: impl FnOnce(&mut K)) -> Self {
+        f(self.get_mut());
+        self
+    }
+
+    pub fn into_object(self) -> K {
+        self.object
+    }
+
+    pub async fn sync(&mut self) -> Result<()>
+    where
+        K: Resource + DeserializeOwned + Serialize + Clone + Debug,
+    {
+        self.object = match self.dirtiness {
+            Dirtiness::New => self.api.create(&PostParams::default(), &self.object).await?,
+            Dirtiness::Dirty => {
+                self.api
+                    .replace(
+                        self.object.meta().name.as_deref().unwrap(),
+                        &PostParams::default(),
+                        &self.object,
+                    )
+                    .await?
```

Comment:
I guess this is going to be the most contentious part as using `::replace` here makes it a bit brittle w.r.t. interacting field managers / whatever admission controllers is battling for the object's rights, while everything in kubernetes is moving to a patch/apply world.

Kind of wondering if it would be useful for json-patches (or apply configurations down the line in #649) to be constructible in some other way from the root structs, and then they could be buffered herein for a bigger patch call. At least that could help control the size / sanity of the api calls so they are less likely to fail / require retries.

Reply:
Do you know what client-go does in this situation? I think they have some implicit syncing going on.

Reply:
We can't use `apply` or `patch` here, because the get-or-create and get-and-modify patterns are both inherently prone to race conditions, and patches bypass the `resourceVersion` check. `apply` is even worse, since for get-and-modify you end up taking ownership of the whole object.

Reply:
Yeah, you are right. I think there's probably two avenues we can take for these helpers; ones that modify objects (using `replace`) that maybe could be retry-wrapped somehow, and ones that consist of combinable sets of json patches (using `patch`).

Incidentally, there was another instance of the first category here propping up in https://github.com/kube-rs/kube-rs/issues/428#issuecomment-1033592469 .

Reply:
I'd rather separate them into "things that require (fallible) server coordination" and "things that manipulate client-side objects".

Yes, users will have to retry when this fails but I think we have to leave this up to them rather than doing this ourselves. Sometimes the `default` function may not be safe to retry, sometimes they have a coarser-grained retry policy that they'd want to bubble up to (such as when running under the `Controller`/`applier`).